### ISSUE TITLE
Code Editor setting for Python IntelliSense in Jupyter notebooks

### ISSUE THREAD

**Issue by SeijiSuenaga** [aws/sagemaker-distribution#319]
When using Code Editor in SageMaker Studio, I've found that Python cells in Jupyter notebooks have no IntelliSense unless I configure this setting, then reload the page:
```json
"python.languageServer": "Jedi"
```

Would it be acceptable to add that to `template/v1/dirs/etc/code-editor/code_editor_machine_settings.json`?

I couldn't find an issue related to this in the microsoft/vscode repository, but did find it mentioned in another repository derived from VS Code: https://github.com/coder/code-server/issues/6323#issuecomment-1806910939

**Comment by edwardps** [comment]
Thanks Seiji for reporting this. We will track this request and evaluate this default languageServer configuration internal.

**Comment by claytonparnell** [comment]
@edwardps are there any updates we can share here?

**Comment by j0bekt01** [comment]
With notebooks, IntelliSense defaults to Jedi. Set it to Pylance instead, as it is much more feature-rich than Jedi. You can add `"python.languageServer": "Pylance"` to your startup lifecycle policy.

**Comment by SeijiSuenaga** [comment]
@j0bekt01 Does that require Pylance to be installed? As far as I know, the Pylance extension isn't available in Code Editor because Microsoft restricts it to their proprietary VS Code distribution.

**Comment by j0bekt01** [comment]
There's a Python extension that's already installed in our SM so I either go to the extension settings and pick Pylance over Jedi or append the line above to the settings file.

**Comment by pjoe** [comment]
I think @SeijiSuenaga is correct. When you select `pylance` code editor prompts to install the extension `ms-python.vscode-pylance`,  but is unable to do so. I tried manually downloading the vsix, but this also fails installing, with an error about vscode version. Also noticed that extension is not on OpenVSX :S

**Comment by pjoe** [comment]
Jedi does enable some limited intellisense for python stdlibs, but nothing for e.g. spark :S

![image](https://github.com/user-attachments/assets/cea8e240-b844-4941-8803-3726fd0f5532)

**Comment by j0bekt01** [comment]
> I think @SeijiSuenaga is correct. When you select `pylance` code editor prompts to install the extension `ms-python.vscode-pylance`, but is unable to do so. I tried manually downloading the vsix, but this also fails installing, with an error about vscode version. Also noticed that extension is not on OpenVSX :S

This extension is on openVSX
https://open-vsx.org/extension/ms-python/python

it should already be baked into the distribution image see here: https://github.com/aws/sagemaker-distribution/blob/main/build_artifacts/v2/v2.0/v2.0.0/dirs/etc/code-editor/extensions.txt

**Comment by pjoe** [comment]
@j0bekt01 this is not the same extension, we are talking about the pylance extension, NOT the python extension, see here:

https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance

vs.

https://marketplace.visualstudio.com/items?itemName=ms-python.python

note the different IDs `ms-python.python` vs. `ms-python.vscode-pylance`

**Comment by j0bekt01** [comment]
> @j0bekt01 this is not the same extension, we are talking about the pylance extension, NOT the python extension, see here:
> 
> https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance
> 
> vs.
> 
> https://marketplace.visualstudio.com/items?itemName=ms-python.python
> 
> note the different IDs `ms-python.python` vs. `ms-python.vscode-pylance`

oh, I see. The openvsx one should have support for pylance

**Comment by pjoe** [comment]
It's not quite like that.

`ms-python.python` has support for automatically **installing** `ms-python.vscode-pylance`. However installing the pylance extension is what fails

**Comment by imarquart** [comment]
pylance does not exist on openvsx and we were not successful activating Pylance in any sort of configuration.
The OpenVSX python extension can not install Pylance, as it does not exist. It fails in installing Pylance, but does so silently. Setting LanguageServer to Pylance then leads to no Python support at all. Jedi is, of course, vastly inferior.

As such:
- Is there any way to remediate this issue?
- Are there any plans to provide better Python support?

If there really isn't any way to do this, or if the AWS team really is not aware of the issue, I also have the following question:
- Has anyone at AWS actually ever tried productive coding in Python using Sagemaker?

**Comment by SeijiSuenaga** [comment]
@imarquart FWIW, for plain .py files you can install the [Pyright](https://open-vsx.org/extension/ms-pyright/pyright) extension, which Pylance is built upon.  It includes some [basic language server features](https://microsoft.github.io/pyright/#/features?id=language-server-support) like renaming variables, but it doesn't work in notebook cells. ☹️

Since Microsoft made Pylance proprietary, I feel like the Python ecosystem needs an open-source alternative that works in notebooks, but I haven't found one (and I doubt I could build one myself).  There is a Pyright fork called [basedpyright](https://docs.basedpyright.com/latest/) that re-implements some Pylance features, but it currently [doesn't support notebooks](https://github.com/DetachHead/basedpyright/issues/173) either.  It'd be awesome if AWS used some of their vast resources to make their own fork for SageMaker... 😉

**Comment by Steinkreis** [comment]
+1 We want to use the Code Editor more extensively at our company in the future, but we definitely need syntax highlighting and proper intelisense configuration. Currently it's just very painful to work with imo.

**Comment by Papeyaa** [comment]
I agree with the others here. We plan to move to Sagemaker Code Editor instead of using local IDEs, since testing pipelines is much easier. But developing there is frustrating without intellisense and proper code highlighting.

**Comment by josh-gree** [comment]
This is an absolute necessity for any productive use of Code Editor!

**Comment by Julfried** [comment]
Would agree with @josh-gree . As far as I understand the issue is Microsoft made Pylance proprietary, which prevents the installation in unofficial builds of VSCode. With that beeing said, the recent developments in [basedpyright](https://github.com/DetachHead/basedpyright) look promising and they even added support for jupyter notebooks in the recent release. I think this is worth looking at and maybe should be included in the future.
